volatility3/cli/volshell/generic.py

```diff
@@ -418,67 +445,194 @@ def display_type(
-                    data_format.byteorder,
-                    "signed" if data_format.signed else "unsigned",
-                )
-            )
+        # add special case for pointer so that information about the struct the
+        # pointer is pointing to is shown rather than simply the fact this is a
+        # pointer object. The "dereference_count < MAX_DEREFERENCE_COUNT" is to
+        # guard against loops
+        dereference_count = 0
+        while (
+            isinstance(volobject, objects.Pointer)
+            and dereference_count < MAX_DEREFERENCE_COUNT
+        ):
+            # before defreerencing the pointer, show it's information
+            print(f'{"    " * dereference_count}{self._display_simple_type(volobject)}')
+
+            # check that we can follow the pointer before dereferencing and do not
+            # attempt to follow null pointers.
+            if volobject.is_readable() and volobject != 0:
+                # now deference the pointer and store this as the new volobject
+                volobject = volobject.dereference()
+                dereference_count = dereference_count + 1
+            else:
+                # if we aren't able to follow the pointers anymore then there will
+                # be no more information to display as we've already printed the
+                # details of this pointer including the fact that we're not able to
+                # follow it anywhere
+                return
 
         if hasattr(volobject.vol, "members"):
+            # display the header for this object, if the orginal object was just a type string, display the type information
+            struct_header = f'{"    " * dereference_count}{volobject.vol.type_name} ({volobject.vol.size} bytes)'
+            if isinstance(object, str) and offset is None:
+                suffix = ":"
+            else:
+                # this is an actual object or an offset was given so the offset should be displayed
+                suffix = f" @ {hex(volobject.vol.offset)}:"
+            print(struct_header + suffix)
+
+            # it is a more complex type, so all members also need information displayed
             longest_member = longest_offset = longest_typename = 0
             for member in volobject.vol.members:
                 relative_offset, member_type = volobject.vol.members[member]
                 longest_member = max(len(member), longest_member)
                 longest_offset = max(len(hex(relative_offset)), longest_offset)
-                longest_typename = max(len(member_type.vol.type_name), longest_typename)
+                member_type_name = self._get_type_name_with_pointer(
+                    member_type
+                )  # special case for pointers to show what they point to
+                longest_typename = max(len(member_type_name), longest_typename)
+                if longest_typename > MAX_TYPENAME_DISPLAY_LENGTH:
+                    longest_typename = MAX_TYPENAME_DISPLAY_LENGTH
```

Comment:
I think `max` can take multiple parameters, so... this should work (although it might need reformatting for ruff/black)?
```suggestion
                longest_typename = max(len(member_type_name), longest_typename, MAX_TYPENAME_DISPLAY_LENGTH)
```

Reply:
The issue with that is then max might be 8000 for the very long type name. 

We don't actually want the max of all three, we want the biggest unless it's reached the cut off.

Reply:
Oh... hmmm... so it's almost like `min(max(len(member_type_name), longest_type_name), MAX_TYPENAME_DISPLAY_LENGTH)` then?

Reply:
Yes - that's exactly right.

Reply:
I've swapped to using min, but put it on another line with some extra comments to hopefully keep readable. It looks better than the if imo.